typing/ctype.ml

```diff
@@ -3259,15 +3259,28 @@ let unify uenv ty1 ty2 =
 
 let unify_gadt (penv : Pattern_env.t) ty1 ty2 =
   let equated_types = TypePairs.create 0 in
-  let uenv = Pattern
-      { penv;
-        equated_types;
-        assume_injective = true;
-        unify_eq_set = TypePairs.create 11; }
-  in
-  with_univar_pairs [] (fun () ->
+  let do_unify_gadt () =
+    let uenv = Pattern
+        { penv;
+          equated_types;
+          assume_injective = true;
+          unify_eq_set = TypePairs.create 11; }
+    in
     unify uenv ty1 ty2;
-    equated_types)
+    equated_types
+  in
+  let closed = closed_type_expr ty1 && closed_type_expr ty2 in
+  if closed then with_univar_pairs [] do_unify_gadt else
```

Comment:
As an experiment selecting gadt unification as soon as `ty2` is closed restores trunk performance for my "knight's tour benchmark with no changes in the testsuite.

Reply:
Would `closed_type_expr ty1 || closed_type_expr ty2` work as well?

The way I understand only checking `ty2` is that we implicit assume that we are checking a more restricted "matching" problem in practice (only the left-hand-side has unification variables), but in practice `the unify_gadt`-using code in typecore.ml does not obviously appear to respect this asymmetric convention.

Reply:
It doesn't work on the example in the testsuite (and the one reported in the previous PRs), since the GADT constructor has a closed type.

Reply:
No, Florian's idea is to consider this as an asymmetric problem: since `unify_gadt` is only called for GADT constructors, and the type variables on the constructor instance side `ty1` are fresh, there is no problem in reifying them as long as they do not leak, and they won't if `ty2` is closed. If we do the same thing when  `ty1` only is closed, then we lose the example we are trying to fix.
The main concern is that we must now distinguish between the two arguments of `unify_gadt`. The calls in `solve_Ppat_construct` respect this asymmetry (with the twist that for some reason `unify_head_only` inverts its parameter order), but in `check_counter_example_pat`, all unifications call `unify_pat_types_refine ~refine:true`, so it is harder to ensure this asymmtery.  Yet, `check_partial` and `check_unused` are both called on `ty_arg_check`, which is a copy of `ty_arg'` in which type variables have been copied too, so they cannot escape either.
Conclusion, it seems that Florian's suggestion is correct, and further, calls from `check_counter_example_pat` do not need to bother about their free type variables.

Reply:
My intuition is that if the function is asymmetric, then its naming should say so, explain what is the desired use-case, and the callsites where it is not-obvious that the use-case is respected should be commented to explain why they are fine. In particular, you should rename `unify_head_only` and swap its parameter order.

Currently you are saying that the code is fine, but this is for a highly non-obvious reason that we will quickly forget, and we may break this invariant in the future. The result will be hard to detect because we don't have good coverage for these cases in our testsuite, and will lead to more confusion later.

Reply:
Actually, `unify_head_only` is unrelated, since it only calls `unify_gadt` when typing counter-examples.
There is one single call-site outside of counter-examples, and it is `unify_res` inside `solve_Ppat_construct`.
I have already added a comment there, and there is a comment in ctype.mli for `unify_gadt`.
As I wrote below, the best way to clarify the code is to clarify the relation between `penv` and calls to `unify_gadt`, but it seems better done in a separate PR (I can of course do it in this one too).